server/src/runner/mod.rs

```diff
@@ -60,38 +64,62 @@ impl Runner {
         };
 
         Ok(Self {
-            docker: docker.context("failed to initialize docker")?
+            docker: docker.context("failed to initialize docker")?,
+            broadcast
         })
     }
 
     /// builds the package inside a container
-    pub async fn build(&self, container: &ContainerId) -> anyhow::Result<RunStatus> {
+    pub async fn build(&self, container: &ContainerId, package: &Package) -> anyhow::Result<RunStatus> {
         let start = Utc::now();
 
         // start container
         self.docker.start_container(container, None::<StartContainerOptions<String>>).await?;
 
-        // wait for container to exit and collect logs
-        let result =
-            self.docker.wait_container(container,  None::<WaitContainerOptions<String>>).collect::<Vec<_>>().await;
-
-        let end = Utc::now();
+        self.broadcast.notify(&package.base, Event::BuildStart).await;
 
         // retrieve logs
         let log_options = LogsOptions {
             stdout: true, stderr: true,
+            follow: true, // follow is needed since we continuously read from the stream
             since: start.timestamp(),
             ..Default::default()
         };
 
-        let logs: Vec<String> = self.docker.logs::<String>(container, Some(log_options)).filter_map(|r| async {
-            r.ok().map(|c| c.to_string())
-        }).collect::<Vec<_>>().await;
+        let mut stream = self.docker.logs::<String>(container, Some(log_options));
+        let base = package.base.clone();
+        let broadcast = self.broadcast.clone();
+        let log_collector = tokio::spawn(async move {
+            let mut logs = vec![];
+            // collect logs from stream until the container exits (and the log stream closes)
+            while let Some(next) = stream.next().await {
```

Comment:
style nit: you could combine this into one if let (with `Some(Ok(log))`), or at least into an if let if we don't want to "exit" when the output returns an error